EIPS/eip-3403.md

```diff
@@ -0,0 +1,138 @@
+
+Gas refunds additionally have multiple harmful consequences:
+
+* Refunds give rise to GasToken. GasToken has benefits in moving gas space from low-fee periods to high-fee periods, but it also has downsides to the network, particularly in exacerbating state size (as state slots are effectively used as a "battery" to save up gas) and inefficiently clogging blockchain gas usage
+* Refunds increase block size variance. The theoretical maximum amount of actual gas consumed in a block is nearly twice the on-paper gas limit (as refunds add gas space for subsequent transactions in a block, though refunds are capped at 50% of a transaction's gas used). This is not fatal, but is still undesirable, especially given that refunds can be used to maintain 2x usage spikes for far longer than EIP 1559 can.
+
+## Specification
+
+### Parameters
+
+| Constant | Value |
+| - | - |
+| `FORK_BLOCK` | TBD |
+| `SSTORE_REFUND_GAS` | 15000 |
+
+For blocks where `block.number >= FORK_BLOCK`, the following changes apply.
+
+1. Remove the `SELFDESTRUCT` refund.
+2. Remove the `SSTORE` refund in all cases except for one specific case: if the _new value_ and _original value_ of the storage slot both equal 0 but the _current value_ does not (those terms being defined as in [EIP-1283](https://eips.ethereum.org/EIPS/eip-1283)), refund `SSTORE_REFUND_GAS` gas.
+
+## Rationale
+
+Preserving refunds in the `new = original = 0 != current` case ensures that a few key use cases that deserve favorable gas cost treatment continue to receive favorable gas cost treatment, particularly:
+
+* Anti-reentrancy locks (typically flipped from 0 to 1 right before a child call begins, and then flipped back to 0 when the child call ends)
+* ERC20 approve-and-send (the "approved value" goes from zero to nonzero when the token transfer is approved, and then back to zero when the token transfer processes)
+
+It also preserves two key goals of EIP 3298:
+
+1. Gas tokens continue to be non-viable, because each 15000 refund is only possible because of 15000 extra gas that was paid for flipping that storage slot from zero to nonzero earlier in the same transaction, so you can't clear some storage slots and use that saved gas to fill others.
+2. The total amount of gas _spent on execution_ is capped at the gas limit. Every 15000 refund for flipping a storage slot non from zero -> zero is only possible because of 15000 extra gas paid for flipping that slot from zero -> nonzero earlier in the same transaction; that gas paid for storage expansion and not execution, and so does not contribute to short-term DoS risk.
```

Comment:
```suggestion
2. The total amount of gas _spent on execution_ is capped at the gas limit. Every 19000 refund for flipping a storage slot non from zero -> zero is only possible because of 19000 extra gas paid for flipping that slot from zero -> nonzero earlier in the same transaction; that gas paid for storage expansion and not execution, and so does not contribute to short-term DoS risk.
```